### PR TITLE
Dwarf: implement .eh_frame

### PR DESCRIPTION
**Pull request by jacobly0**

### CODE REVIEW

File: src/link/Elf.zig [hunk 1]
```diff
@@ -3104,43 +3107,81 @@ fn initSyntheticSections(self: *Elf) !void {
...
+                        const index = try zig_object.newSymbolWithAtom(alloc, name_off);
+                        const sym = zig_object.symbol(index);
+                        const esym = &zig_object.symtab.items(.elf_sym)[sym.esym_index];
+                        esym.st_info |= elf.STT_SECTION;
+                        const atom_ptr = zig_object.atom(sym.ref.index).?;
+                        atom_ptr.alignment = alignment;
+                        atom_ptr.output_section_index = shndx;
+                        return index;
+                    }
+                }.addSectionSymbol;
```
Comment (kubkon) [inline src/link/Elf.zig:3141]: We already have this logic defined and used in `initMetadata`. Again, we should not call `initSyntheticSections` from `initMetadata`.

File: src/link/Elf/relocatable.zig [hunk 2]
```diff
@@ -289,6 +289,8 @@ fn claimUnresolved(elf_file: *Elf) void {
 }
 
 fn initSections(elf_file: *Elf) !void {
+    const comp = elf_file.base.comp;
+    const target = comp.root_mod.resolved_target.result;
```
Comment (kubkon) [inline src/link/Elf/relocatable.zig:293]:
```suggestion
    const target = elf_file.getTarget();
```

**Review by kubkon — CHANGES_REQUESTED** [review]
Perhaps it would make sense to land `elf-incr` branch first in its current state since it cleans up a lot of things in the ELF linker, and then rebase this branch on top? Next, `initSyntheticSections` really should not be used in `initMetadata` - happy to clean that one up if you want. Finally, not a blocker but as it currently stands, `.eh_frame` generated with `ZigObject` will not have its CIEs deduped across all input object files - not a biggie, but something to remember about. I would also like to have it tested with the linker - emit `.eh_frame` section as part of `.o` file and then link it back into an executable.

**Comment by jacobly0** [comment]
> Finally, not a blocker but as it currently stands, `.eh_frame` generated with `ZigObject` will not have its CIEs deduped across all input object files - not a biggie, but something to remember about. I would also like to have it tested with the linker - emit `.eh_frame` section as part of `.o` file and then link it back into an executable.

A CIE is like 24 bytes, so it's definitely not worth deduping the two extra CIE per module in incremental mode (the only use case I'm working towards right now), especially given all the padding that already exists.  The key feature that would be useful is the ability to generate a `.eh_frame_hdr` in non-incremental mode, which I believe would come for free when CIE deduping is integrated.

**Review by kubkon — APPROVED** [review]
Looks great!

**Review by mlugg — APPROVED** [review]

**Pull request merged**